File: references/java-sdk/src/main/java/blue/language/processor/DocumentProcessor.java
package blue.language.processor;

import blue.language.mapping.NodeToObjectConverter;
import blue.language.model.Node;
import blue.language.processor.model.Contract;
import blue.language.processor.model.MarkerContract;
import blue.language.utils.TypeClassResolver;
import java.util.Map;
import java.util.Objects;

/**
 * Facade over the processor engine; retains public API for Document processing.
 */
public class DocumentProcessor {

    private static final TypeClassResolver CONTRACT_TYPE_RESOLVER =
            new TypeClassResolver("blue.language.processor.model");

    private final ContractProcessorRegistry contractRegistry;
    private final NodeToObjectConverter contractConverter;
    private final ContractLoader contractLoader;
    private final TypeClassResolver contractTypeResolver;

    public DocumentProcessor() {
        this(ContractProcessorRegistryBuilder.create().registerDefaults().build(), CONTRACT_TYPE_RESOLVER);
    }

    public DocumentProcessor(ContractProcessorRegistry registry) {
        this(registry, CONTRACT_TYPE_RESOLVER);
    }

    public DocumentProcessor(ContractProcessorRegistry registry, TypeClassResolver resolver) {
        this.contractRegistry = Objects.requireNonNull(registry, "registry");
        this.contractTypeResolver = resolver != null ? resolver : CONTRACT_TYPE_RESOLVER;
        this.contractRegistry
                .nodeProvider(this.contractTypeResolver.nodeProvider())
                .typeClassResolver(this.contractTypeResolver);
        this.contractConverter = new NodeToObjectConverter(this.contractTypeResolver);
        this.contractLoader = new ContractLoader(contractRegistry, contractConverter);
    }

    private DocumentProcessor(Builder builder) {
        this(builder.contractRegistry, builder.typeClassResolver);
    }

    public DocumentProcessingResult initializeDocument(Node document) {
        return ProcessorEngine.initializeDocument(this, document);
    }

    public DocumentProcessingResult processDocument(Node document, Node event) {
        return ProcessorEngine.processDocument(this, document, event);
    }

    public boolean isInitialized(Node document) {
        return ProcessorEngine.isInitialized(this, document);
    }

    public DocumentProcessor registerContractProcessor(ContractProcessor<? extends Contract> processor) {
        contractRegistry.register(processor);
        return this;
    }

    public ContractProcessorRegistry getContractRegistry() {
        return contractRegistry;
    }

    ContractProcessorRegistry registry() {
        return contractRegistry;
    }

    NodeToObjectConverter contractConverter() {
        return contractConverter;
    }

    ContractLoader contractLoader() {
        return contractLoader;
    }

    public Map<String, MarkerContract> markersFor(Node scopeNode, String scopePath) {
        ContractBundle bundle = contractLoader.load(scopeNode, scopePath);
        return bundle.markers();
    }

    public static Builder builder() {
        return new Builder();
    }

    static final class Builder {
        private ContractProcessorRegistry contractRegistry = ContractProcessorRegistryBuilder.create().registerDefaults().build();
        private TypeClassResolver typeClassResolver = CONTRACT_TYPE_RESOLVER;

        public Builder withRegistry(ContractProcessorRegistry registry) {
            this.contractRegistry = Objects.requireNonNull(registry, "registry");
            return this;
        }

        public Builder withTypeClassResolver(TypeClassResolver resolver) {
            this.typeClassResolver = Objects.requireNonNull(resolver, "resolver");
            return this;
        }

        public DocumentProcessor build() {
            return new DocumentProcessor(this);
        }
    }
}

File: libs/document-processor/src/api/document-processor.ts
import { Blue, BlueNode } from '@blue-labs/language';
import type { BexEngine } from '@blue-labs/bex';

import { ContractLoader } from '../engine/contract-loader.js';
import { createDefaultMergingProcessor } from '../merge/utils/default.js';
import { ProcessorEngine } from '../engine/processor-engine.js';
import type { ProcessorRuntimeHooks } from '../engine/processor-engine.js';
import type { MarkerContract } from '../model/index.js';
import { ContractProcessorRegistry } from '../registry/contract-processor-registry.js';
import { ContractProcessorRegistryBuilder } from '../registry/contract-processor-registry-builder.js';
import type { AnyContractProcessor } from '../registry/types.js';
import { blueRepository } from '../repository/semantic-repository.js';
import type { DocumentProcessingResult } from '../types/document-processing-result.js';

const DEFAULT_BLUE = new Blue({
  repositories: [blueRepository],
  mergingProcessor: createDefaultMergingProcessor(),
});

export interface DocumentProcessorOptions {
  readonly blue?: Blue;
  readonly bexEngine?: BexEngine;
  readonly registry?: ContractProcessorRegistry;
  readonly runtimeHooks?: ProcessorRuntimeHooks;
}

export class DocumentProcessor {
  private readonly blue: Blue;
  private readonly registryRef: ContractProcessorRegistry;
  private readonly contractLoaderRef: ContractLoader;
  private readonly engine: ProcessorEngine;

  constructor(options?: DocumentProcessorOptions) {
    this.registryRef =
      options?.registry ??
      ContractProcessorRegistryBuilder.create({
        bexEngine: options?.bexEngine,
      })
        .registerDefaults()
        .build();
    this.blue = options?.blue ?? DEFAULT_BLUE;
    this.contractLoaderRef = new ContractLoader(this.registryRef, this.blue);
    this.engine = new ProcessorEngine(
      this.contractLoaderRef,
      this.registryRef,
      this.blue,
      options?.runtimeHooks,
    );
  }

  registerContractProcessor(processor: AnyContractProcessor): this {
    this.registryRef.register(processor);
    return this;
  }

  async initializeDocument(
    document: BlueNode,
  ): Promise<DocumentProcessingResult> {
    return this.engine.initializeDocument(document);
  }

  async processDocument(
    document: BlueNode,
    event: BlueNode,
  ): Promise<DocumentProcessingResult> {
    return this.engine.processDocument(document, event);
  }

  markersFor(
    scopeNode: BlueNode,
    scopePath: string,
  ): Map<string, MarkerContract> {
    const bundle = this.contractLoaderRef.load(scopeNode, scopePath);
    return bundle.markers();
  }

  isInitialized(document: BlueNode): boolean {
    return this.engine.isInitialized(document);
  }

  getContractRegistry(): ContractProcessorRegistry {
    return this.registryRef;
  }

  /** @internal */
  registry(): ContractProcessorRegistry {
    return this.registryRef;
  }

  /** @internal */
  contractLoader(): ContractLoader {
    return this.contractLoaderRef;
  }

  static builder(): DocumentProcessorBuilder {
    return new DocumentProcessorBuilder();
  }
}

export class DocumentProcessorBuilder {
  private contractRegistry: ContractProcessorRegistry;
  private blueInstance: Blue | undefined;
  private bexEngine: BexEngine | undefined;

  constructor() {
    this.contractRegistry = ContractProcessorRegistryBuilder.create()
      .registerDefaults()
      .build();
  }

  withRegistry(registry: ContractProcessorRegistry): DocumentProcessorBuilder {
    this.contractRegistry = registry;
    return this;
  }

  registerDefaults(): DocumentProcessorBuilder {
    return this;
  }

  withBlue(blue: Blue): DocumentProcessorBuilder {
    this.blueInstance = blue;
    return this;
  }

  withBexEngine(bexEngine: BexEngine): DocumentProcessorBuilder {
    this.bexEngine = bexEngine;
    this.contractRegistry = ContractProcessorRegistryBuilder.create({
      bexEngine,
    })
      .registerDefaults()
      .build();
    return this;
  }

  build(): DocumentProcessor {
    return new DocumentProcessor({
      registry: this.contractRegistry,
      blue: this.blueInstance,
      bexEngine: this.bexEngine,
    });
  }
}
